From 3a4a23828b6a0487f26e6b86672f8fa81e4666bc Mon Sep 17 00:00:00 2001
From: "Jiang, Renzhi" <renzhi.jiang@intel.com>
Date: Thu, 22 Sep 2022 13:13:39 +0800
Subject: [PATCH 09/17] remove debug information (#11)

---
 build.sh | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/build.sh b/build.sh
index fa5ff5d06e..31420303ad 100755
--- a/build.sh
+++ b/build.sh
@@ -26,9 +26,9 @@ cmake \
 -DENABLE_FUNCTIONAL_TESTS=OFF \
 -DENABLE_OV_CORE_UNIT_TESTS=OFF \
 -DENABLE_OV_CORE_BACKEND_UNIT_TESTS=OFF \
--DENABLE_DEBUG_CAPS=ON \
--DENABLE_GPU_DEBUG_CAPS=ON \
--DENABLE_CPU_DEBUG_CAPS=ON \
+#-DENABLE_DEBUG_CAPS=ON \
+#-DENABLE_GPU_DEBUG_CAPS=ON \
+#-DENABLE_CPU_DEBUG_CAPS=ON \
 -DCMAKE_BUILD_TYPE=$1 \
  ..
 
-- 
2.25.1

